Sources/TuistGenerator/Generator/DescriptorGenerator.swift

```diff
@@ -0,0 +1,85 @@
+import Basic
+import Foundation
+import TuistCore
+import TuistSupport
+import XcodeProj
+
+public struct GeneratedSchemeDescriptor {
+    public var scheme: XCScheme
+    public var shared: Bool
+}
+
+public enum GeneratedSideEffect {
+    case file(GeneratedFile)
+    case delete(AbsolutePath)
+    case command(GeneratedCommand)
+}
+
+public struct GeneratedFile {
+    public var path: AbsolutePath
+    public var contents: Data
+}
+
+public struct GeneratedCommand {
+    public var command: [String]
+}
+
+public struct GeneratedProjectDescriptor {
+    public var path: AbsolutePath
+    public var xcodeProj: XcodeProj
+    public var schemes: [GeneratedSchemeDescriptor]
+    public var sideEffects: [GeneratedSideEffect]
+}
+
+public struct GeneratedWorkspaceDescriptor {
+    public var path: AbsolutePath
+    public var xcworkspace: XCWorkspace
+    public var projects: [GeneratedProjectDescriptor]
+    public var schemes: [GeneratedSchemeDescriptor]
+    public var sideEffects: [GeneratedSideEffect]
+}
+
+/// Descriptor Generator
+///
+/// Produces a side effect free representation of a generated project or workspace
+///
+///
+public protocol DescriptorGenerating {
```

Comment:
@kwridan I have to say this abstraction is very beautiful. I really like the direction that we are taking here.